src/libcrun/cgroup.c

```diff
@@ -330,6 +330,87 @@ is_rootless (libcrun_error_t *err)
   return check_running_in_user_namespace (err);
 }
 
+static int
+get_file_owner (const char *path, uid_t *uid, gid_t *gid)
+{
+  struct stat st;
+  int ret;
+
+#ifdef HAVE_STATX
+  struct statx stx;
+
+  ret = statx (AT_FDCWD, path, AT_STATX_DONT_SYNC, STATX_UID|STATX_GID, &stx);
+  if (UNLIKELY (ret < 0))
+    {
+      if (errno == ENOSYS)
+        goto fallback;
+
+      return ret;
+    }
+  *uid = stx.stx_uid;
+  *gid = stx.stx_gid;
+  return ret;
+
+ fallback:
+#endif
+  ret = stat (path, &st);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  *uid = st.st_uid;
+  *gid = st.st_gid;
+  return ret;
+}
+static int
+chown_cgroups (const char *path, uid_t uid, gid_t gid, libcrun_error_t *err)
+{
+  cleanup_free char *cgroup_path = NULL;
+  cleanup_dir DIR *dir = NULL;
+  struct dirent *next;
+  int ret;
+  int dfd;
+
+  xasprintf (&cgroup_path, "/sys/fs/cgroup/%s", path);
+
+  dir = opendir (cgroup_path);
+  if (UNLIKELY (dir == NULL))
+    return crun_make_error (err, errno, "cannot opendir %s", cgroup_path);
+
+  dfd = dirfd (dir);
+
+  for (next = readdir (dir); next; next = readdir (dir))
+    {
+      const char *name = next->d_name;
+
+      /* do not chown the parent directory, but chown the current one.  */
+      if (name[0] == '.' && name[1] == '.')
+        continue;
+
+      ret = fchownat (dfd, name, uid, gid, AT_SYMLINK_NOFOLLOW);
```

Comment:
Should you check if the `to` already has the correct UID/GID, before doing the fchownat?

Reply:
at this point we are sure the files are owned by root in the user namespace, as the cgroup was created by the caller of this function.  The caller also makes sure to not attempt the chown'ing if root in the user namespace is the same user that created the cgroup (https://github.com/containers/crun/pull/123/files#diff-708d0d5dba5aca61dc787ceee8bc3a6cR1329-R1330)